editor/assets/default_renderpipeline/builtin-pipeline-types.ts

```diff
@@ -32,313 +32,159 @@ import { Material, Texture2D, gfx } from 'cc';
+    sharpness = 0.8;
 }
 
 export function fillRequiredFSR(value: FSR): void {
-    if (value.enabled === undefined) {
-        value.enabled = false;
-    }
-    if (value.material === undefined) {
-        value.material = null;
-    }
-    if (value.sharpness === undefined) {
-        value.sharpness = 0.8;
-    }
-}
-
-export interface FXAA {
-    enabled: boolean; /* false */
-    /* refcount */ material: Material | null;
-    [name: string]: unknown;
+    value.enabled ??= false;
+    value.material ??= null;
+    value.sharpness ??= 0.8;
 }
 
-export function makeFXAA(): FXAA {
-    return {
-        enabled: false,
-        material: null,
-    };
+export class FXAA {
+    enabled = false;
+    /* refcount */ material: Material | null = null;
 }
 
 export function fillRequiredFXAA(value: FXAA): void {
-    if (value.enabled === undefined) {
-        value.enabled = false;
-    }
-    if (value.material === undefined) {
-        value.material = null;
-    }
-}
-
-export interface ToneMapping {
-    /* refcount */ material: Material | null;
-    [name: string]: unknown;
+    value.enabled ??= false;
+    value.material ??= null;
 }
 
-export function makeToneMapping(): ToneMapping {
-    return {
-        material: null,
-    };
+export class ToneMapping {
+    /* refcount */ material: Material | null = null;
 }
 
 export function fillRequiredToneMapping(value: ToneMapping): void {
-    if (value.material === undefined) {
-        value.material = null;
-    }
-}
-
-export interface PipelineSettings {
-    readonly msaa: MSAA;
-    enableShadingScale: boolean; /* false */
-    shadingScale: number; /* 0.5 */
-    readonly depthOfField: DepthOfField;
-    readonly bloom: Bloom;
-    readonly toneMapping: ToneMapping;
-    readonly colorGrading: ColorGrading;
-    readonly fsr: FSR;
-    readonly fxaa: FXAA;
-    [name: string]: unknown;
+    value.material ??= null;
 }
 
-export function makePipelineSettings(): PipelineSettings {
-    return {
-        msaa: makeMSAA(),
-        enableShadingScale: false,
-        shadingScale: 0.5,
-        depthOfField: makeDepthOfField(),
-        bloom: makeBloom(),
-        toneMapping: makeToneMapping(),
-        colorGrading: makeColorGrading(),
-        fsr: makeFSR(),
-        fxaa: makeFXAA(),
-    };
+export class PipelineSettings {
+    readonly msaa: MSAA = new MSAA();
+    enableShadingScale = false;
+    shadingScale = 0.5;
+    readonly depthOfField: DepthOfField = new DepthOfField();
+    readonly bloom: Bloom = new Bloom();
+    readonly toneMapping: ToneMapping = new ToneMapping();
+    readonly colorGrading: ColorGrading = new ColorGrading();
+    readonly fsr: FSR = new FSR();
+    readonly fxaa: FXAA = new FXAA();
 }
 
 export function fillRequiredPipelineSettings(value: PipelineSettings): void {
-    if (value.msaa === undefined) {
-        (value.msaa as MSAA) = makeMSAA();
+    if (!value.msaa) {
+        (value.msaa as MSAA) = new MSAA();
```

Comment:
Why need to convert it to `MSAA` explictly?

Reply:
`value.msaa` is `readonly`, we cannot assign without the cast.

Reply:
Got it.